### PR TITLE
Fix the condition for calling update_learning_rates

### LINKED ISSUE #6616: Epoch level scheduler is not well called except check_val_every_n_epoch == 1
## Please reproduce using the BoringModel

Run the following colab link. 
You can see multi-step lr scheduler is only called at validation epoch (i.e. 4 - check_val_every_epoch is set as 4) even I set milestones as [1,2,3,4]. I set scheduler attribute as verbose=True to show changed lr_rates. 
If check_val_every_epoch is set as 1, scheduler "setup()" after train epoch seems well called. 

https://colab.research.google.com/drive/1IqgyIU6EsCTk1y1_xH6VIEdpDsoMyPyt?usp=sharing

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

Result : 

GPU available: True, used: False
TPU available: None, using: 0 TPU cores
/usr/local/lib/python3.7/dist-packages/pytorch_lightning/utilities/distributed.py:51: UserWarning: GPU available but not used. Set the --gpus flag when calling the script.
  warnings.warn(*args, **kwargs)
**_Adjusting learning rate of group 0 to 1.0000e-01._** --> **Initial lr set verbose**

  | Name  | Type   | Params
---------------------------------
0 | layer | Linear | 66    
---------------------------------
66        Trainable params
0         Non-trainable params
66        Total params
0.000     Total estimated model params size (MB)
Validation sanity check: 0%
0/2 [00:00<?, ?it/s]
Epoch 4: 48%
300/626 [00:00<00:00, 374.57it/s, loss=3.06e-14, v_num=0]
Validating:
0/0 [00:00<?, ?it/s]
**_Adjusting learning rate of group 0 to 1.0000e-02._** --> **Incorrect outcome - should be called at every epoch since milestones are set as [1,2,3,4]** 

Testing: 100%
313/313 [00:00<00:00, 777.17it/s]

--------------------------------------------------------------------------------
DATALOADER:0 TEST RESULTS
{'fake_test_acc': 3.062137267926049e-14}
--------------------------------------------------------------------------------

### Expected behavior

Should lr scheulder step is called after train epoch end if interval is set as epoch 

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```
* CUDA:
	- GPU:
		- Tesla K80
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.0+cu101
	- pytorch-lightning: 1.2.4
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by tkng**
## What does this PR do?

This pull request fixes the condition for `update_learning_rates` to be called in `training_loop.py`.

`update_learning_rates` in the `OptimizerConnector` class is called from several places. In the `training_loop.py`, one of the calling conditions is missing a `not`.

As a result, when Trainer's `check_val_every_n_epoch` is set to a value greater than 2, the learning rate is not updated as expected. For example, if `check_val_every_n_epoch` is set to 2, the learning rate will be updated only once every 2 epochs.

### more detail

The current condition for calling `update_learning_rates` is as follows.

```
if (val_loop_called and not should_check_val) or should_train_only
```

The comment above the line says "_update epoch level lr_schedulers if no val loop outside train loop is triggered_". Current code doesn't work as the comment expects.

This pull request fixes this issue by inserting `not` before the `val_loop_called`.

Fixes #6616

By the way, we can assume that `update_learning_rates` does not have any side effects (in fact, I check the current behavior of PyTorch-Lightning by writing a custom learning rate rule by myself. Currently, it is called twice in one epoch). So, I think we can call it every time in the training loop without thinking anything difficult. However, since I don't understand the whole code of this project, I kept the amount of changes in this pull request as small as possible.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Comment by carmocca** [comment]
Can you write a test that fails on current master without the fix?

**Comment by tkng** [comment]
I'm sory that I have no idea how to write a test for this issue...

Even if `update_learning_rates` is not called, training itself will proceed. Also, I can't think of a good way to check that `update_learning_rates` is not being called, from outside the Trainer.

**Comment by awaelchli** [comment]
@tkng You must have observed this bug somehow in your code right? If you provide a minimal example (see our bug_report_model), we can easily help you convert it to a good regression test :)

**Comment by tkng** [comment]
@awaelchli thanks, I wrote a minimal example and uploaded to https://gist.github.com/tkng/a89ffdab2558aedc04a67f7142e6aeb0. (Sorry, I noticed to bug_report_model.py after I wrote this example, so the code is based on lightning-flash.)

Once you run this example and opened the log directory by tensorboard, you'll see this learning rate decay. Pink one is with current pytorch-lightning, green one is with the patch attached on this PR.

**Comment by awaelchli** [comment]
Thanks. Based on your description and code, I pushed a minimal test that fails on master.

**Review by ananthsub — APPROVED** [review]

**Comment by ananthsub** [comment]
Why is the learning rate update dependent on the validation loop at all?

**Comment by awaelchli** [comment]
I suspect because of this scheduler
https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.ReduceLROnPlateau
But I'm not an expert on this section of the code. I believe @rohitgr7 has worked a bit here if I'm not mistaken.

**Review by Borda — APPROVED** [review]
lgtm

**Comment by rohitgr7** [comment]
well, the solution here is a bit incorrect. val_loop_called actually tells whether the val_loop is called within the training loop or not for the cases where val_check_interval(float) < 1. or val_check_interval < num_train_batches. In such a case val_loop_called will be True and thus we need to ensure to explicitly call the epoch level schedulers explicitly outside the train loop if there is no epoch end validation going to happen. I have updated a test https://github.com/PyTorchLightning/pytorch-lightning/pull/7084 which will fail with your changes.

Epoch level schedulers also update within the evaluation loop because we need to ensure they get updated after the validation loop and right before checkpointing to ensure the correct scheduler state dict being saved in the checkpoint file. I know this is a bit complicated but most of the complexity is due to this ReduceLROnPlateau scheduler which needs a monitor to update itself. This monitor can belong to anything logged in either train or val.

To handle the cases where check_val_every_n_epoch > 1, I think we need to somehow pull out the scheduler update from val_loop and still need to ensure that correct scheduler state_dict gets updated within the checkpoint file (we have tests for that) and ofcourse should work with ReduceLROnPlateau (I think we have tests for that too).

**Comment by tkng** [comment]
@rohitgr7 Hmm, the problem seems more complicated than I had thought...

May I assume that update_learning_rates should always be called at the end of one training epoch, even for ReduceLROnPlateau?

If the assumption is correct, then I think the correct code would be to call update_learning_rate immediately after the end of the training process for one epoch, in the run_training_epoch method of TrainLoop class. 

I'm thinking this understanding is consistent with your statement, _I think we need to somehow pull out the scheduler update from val_loop and still need to ensure that correct scheduler state_dict gets updated within the checkpoint file_.

I tried to write such a code once, but I could not make the unit test pass. On the other hand, if I discarded the above understanding and just considered passing the unit test, I only needed to change one line, as I just pushed. (Just for sure, this also passes the test for #7084.) However, I am not sure if the current code is correct. :thinking:

**Comment by rohitgr7** [comment]
yeap `val_loop_called` is redundant since `should_check_val` can cover up that case. Calling lr_update right after the train loop will create problem with `ReduceLROnPlateau` since the monitor might not be logged by then. Also even if we do it after the validation loop then checkpoint won't be getting the correct lr_state_dict. I think I have a solution for this. Will try it on weekend.

**Comment by carmocca** [comment]
Converting to draft so merging is blocked until these comments are addressed 👍

**Comment by tkng** [comment]
@Borda This pull request passes the current all unit tests, but according to @rohitgr7, it is incompatible with ReduceLROnPlateau.

I have tried to rewrite this PR to compatible with ReduceLROnPlateau, but I have failed to write such code that passes all tests and still conforms to ReduceLROnPlateau. I think we have two options.

1. wait for new code from @rohitgr7
2. merge this pull request, since it passes all the current unit tests

I am OK with either, but we'll need a opinion from @rohitgr7.

**Comment by rohitgr7** [comment]
I tried something, but couldn't find an optimal solution that can cover all the cases because of this `ReduceLROnPlateau`. The problem is if a monitor is used with this scheduler with `check_val_every_n_epoch > 1` then there is no way to decide whether to do a `scheduler.step` after validation or after every train loop because this monitor can be anything and can be logged anywhere. For eg `check_val_every_n_epoch>1`, if it's logged in val loop then we ideally shouldn't do `scheduler.step` after every epoch but if it's logged in train loop then we should. Although one can easily configure this correctly by setting `scheduler['interval']` accordingly if they understand how they have configured their train/val loop using trainer parameters.

So I'd say merge this one.

**Review by rohitgr7 — APPROVED** [review]

File: tests/trainer/optimization/test_optimizers.py [hunk 1]
```diff
@@ -600,3 +602,21 @@ def test_optimizer_state_on_device(tmpdir):
...
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=2,
+        limit_val_batches=2,
+        check_val_every_n_epoch=check_val_every_n_epoch,
+        max_epochs=epochs,
+    )
+    trainer.fit(model)
+    assert mocked_sched.call_count == expected_steps
```
Comment (carmocca) [inline tests/trainer/optimization/test_optimizers.py:622]: noob question: why do we want to update schedulers all epochs even if we only run validation some of them?

Comment (rohitgr7) [reply tests/trainer/optimization/test_optimizers.py:622]: running validation on some epoch is just a choice. Ideally, a scheduler should update after every epoch if `scheduler['frequency] = 1(default)` and `scheduler['interval'] == 'epoch'`. If someone wants it to align it with validation, they can set `scheduler['frequency'] = check_val_every_n_epoch`.
